docs/source/contributing.generator.rst

```diff
@@ -0,0 +1,255 @@
+
+Sometimes, we can leave the generator constructor alone and just inherit from ``base.Generator``. In the case of replicate, though, we want to check that there's a Replicate API key in place, and fail early if it's missing. Replicate calls require a user to add an API key, and garak won't be able to do a run without that key - so the polite thing to do is fail as early as we can. Generator load seems like a fine place to do that. The parent class' constructor already manages tracking this value and storing it in ``self.name``.
+
+Thinking more about user experience - when is a good time to quit because of a missing key? If we quit quietly before the module loads, it might be unclear to the user why. So, we first print a progress message about trying to load ``ReplicateGenerator``, and then afterwards check the key. This message printing is handled by the parent class.
+
+So, in the constructor, we first call the parent constructor using ``super().__init__()``, and then do a check for the API key. If the key is missing, we should print a clear message to the user, showing them what the key might look like, and where it should go. And we draw attention to that helpful message with a clear emoji.
+
+.. code-block:: python
+
+        def __init__(self, name, generations=10):
+            super().__init__(name, generations=generations)
+
+            if os.getenv("REPLICATE_API_TOKEN", default=None) is None:
+                raise ValueError(
+                    '🛑 Put the Replicate API token in the REPLICATE_API_TOKEN environment variable (this was empty)\n \
+                    e.g.: export REPLICATE_API_TOKEN="r8-123XXXXXXXXXXXX"'
+                )
+
+The API key is stored in an environment variable, ``REPLICATE_API_TOKEN``, which is accessed using ``os.getenv()``. So don't forget to import that at the top of the module!
+
+.. code-block:: python
+
+    import os
+
+Finally, if the key check passed, let's try to load up the Replicate API using the ``replicate`` module and the user-supplied key. We don't want to do speculative loading in garak - everything should be imported as late as reasonable, to keep user experience fast. So in this case, we import the ``replicate`` API module after the initial validation. Finally, to give the module some persistence, it's loaded at the level of our generator module, instead of just in this method. We add this to the end of ``__init__()``:
+
+.. code-block:: python
+
+            self.replicate = importlib.import_module("replicate")
+
+Finally, forget to import ``importlib`` at the top!
+
+.. code-block:: python
+
+    import importlib
```

Comment:
While I agree on late loading, providing guidance on having `__init__` import of objects required for the class to function is not going to be `pickle` friendly for any multiprocessing.

Reply:
Do we have a threadsafe pattern, e.g. the one in `openai.py`? Shall we move the import to module-level for simplicity?

Reply:
Short term I can live with this guidance, however I suspect we should keep track of this as something we want to iterate on, and remembering there is guidance here may be easily overlooked.

Not sure what the best pattern will be, I think module or class level is easier on the contributor, however the plugin loader currently would take the full import load at that point and remove most of the cost saved by late loading.

If we shift the burden of enumerating the plugin list to an automation job that maintains a cache file used for initial load and only attempt to `search` for plugins that are not `known` to the cache we might be able to shift the import load until the class is actually instantiated during a run for users that are not adding custom plugins.

Reply:
I like the idea of a cache, let's aim for that.